src/tss2-tcti/tcti-device.c

```diff
@@ -136,63 +154,59 @@ tcti_device_receive (
+        LOG_DEBUG ("Caller queried for size but linux kernel doesn't allow this. "
+                   "Returning 4k which is the max size for a response buffer.");
+        *response_size = 4096;
         return TSS2_RC_SUCCESS;
     }
-    if (*response_size < tcti_common->header.size) {
-        LOG_WARNING ("Size of user supplied response buffer %zd is less than "
-                     "the size of the response buffer: %" PRIu32,
-                     *response_size, tcti_common->header.size);
-        return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
-    }
-    *response_size = tcti_common->header.size;
-
-    rc = header_marshal (&tcti_common->header, response_buffer);
-    if (rc != TSS2_RC_SUCCESS) {
-        return rc;
+    if (*response_size < 4096) {
+        LOG_INFO ("Caller provided buffer that *may* not be large enough to "
+                  "hold the response buffer.");
     }
-    /* Read the rest of the response, minus the header that we already jave. */
-    size = read_all (tcti_dev->fd,
-                     &response_buffer [TPM_HEADER_SIZE],
-                     tcti_common->header.size - TPM_HEADER_SIZE);
+    /*
+     * The kernel driver will only return a response buffer in a single read
+     * operation. If we try to read again before sending another command
+     * the kernel will close the file descriptor and we'll get an EOF.
+     */
+    size = TEMP_RETRY (read (tcti_dev->fd, response_buffer, *response_size));
     if (size < 0) {
-        LOG_WARNING ("Failed to read response body.");
-        rc = TSS2_TCTI_RC_IO_ERROR;
+        LOG_ERROR ("Failed to read response from fd %d, got errno %d: %s",
+                   tcti_dev->fd, errno, strerror (errno));
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+    if (size == 0) {
+        LOG_WARNING ("Got EOF instead of response.");
+        rc = TSS2_TCTI_RC_NO_CONNECTION;
         goto out;
     }
-
-    LOGBLOB_DEBUG(response_buffer, tcti_common->header.size, "Response Received");
-
+    LOGBLOB_DEBUG(response_buffer, size, "Response Received");
+    if (size < (ssize_t)TPM_HEADER_SIZE) {
+        LOG_ERROR ("Received %zu bytes, not enough to hold a TPM2 response "
+                   "header. Something is wrong.", size);
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE;
+        goto out;
+    }
+    rc = header_unmarshal (response_buffer, &tcti_common->header);
+    if (rc != TSS2_RC_SUCCESS) {
+        goto out;
+    }
+    if ((size_t)size != tcti_common->header.size) {
+        LOG_INFO ("TPM2 header size disagrees with number of bytes read from "
+                  "fd %d. Header says %u but we read %zu bytes.",
+                  tcti_dev->fd, tcti_common->header.size, *response_size);
+    }
+    if (*response_size < tcti_common->header.size) {
+        LOG_WARNING ("TPM2 response header size is larger than the provided "
+                     "buffer: future use of this TCTI will likely fail.");
+    }
```

Comment:
We should `rc = TSS2_TCTI_RC_GENERAL_FAILURE;` in this case...

Reply:
done